2022-09-28-introduce-q-d.md

```diff
@@ -0,0 +1,241 @@
+    NewTx :: Party -> Payment -> Action WorldState ()
+    Wait :: DiffTime -> Action WorldState ()
+    ObserveConfirmedTx :: Payment -> Action WorldState ()
+```
+
+Then one needs to define:
+* An `initialState`,
+* How to generate `arbitraryAction`s which will be used to produce sequences (or traces) of `Action`s to execute, depending on the current state,
+* A `precondition` function ensuring some `Action` is valid in some state. This function may seem redundant with the generator but is actually important when _shrinking_ a failing test sequences: The execution engine will ensure the reduced sequence is still valid with respect to the model,
+* A `nextState` (transition) function that evolves the model state according to the `Action`s,
+* Auxiliary function `actionName` to providea  human-readable representation of actions.
+
+The reader is invited to check the [Haddock](https://hackage.haskell.org/package/quickcheck-dynamic-1.1.0/docs/Test-QuickCheck-StateModel.html) documentation of the library for further details.
+
+### Exercising Implementation
+
+A _Model_ alone is somewhat useless if we don't provide a way to relate it to the actual implementation of the system-under-test. `quickcheck-dynamic` provides the [`RunModel`](https://hackage.haskell.org/package/quickcheck-dynamic-1.1.0/docs/Test-QuickCheck-StateModel.html#t:RunModel) typeclass for this purpose. The most important function to define is `perform` which defines how `StateModel`'s `Action` should be executed against the implementation within some monadic context `m`. Having the actual execution `Monad m` be a parameter of the `RunModel` gives more flexibility to the implementor which is not tied to `IO` for example.
+
+In the case of Hydra, the `perform` function is defined as:
+
+```haskell
+  perform ::
+    WorldState ->
+    Action WorldState a ->
+    LookUp (StateT (Nodes m) m ->
+    StateT (Nodes m) m a
+  perform st command _ = do
+    case command of
+      Seed{seedKeys} ->
+        seedWorld seedKeys
+      Commit party utxo ->
+        performCommit (snd <$> hydraParties st) party utxo
+...
+```
+
+The actual monad used is a classical `State` monad whose state maps a `Party` to the corresponding client connection to the Hydra node:
+
+```haskell
+data Nodes m = Nodes
+  { nodes :: Map.Map Party (TestHydraNode Tx m)
+  , logger :: Tracer m (HydraNodeLog Tx)
+  }
+```
+
+The `m` parameter is here kept somewhat unconstrained in order to make it possible to run properties within the `IOSim` monad for faster tests execution. Also note the presence of the `logger` field which is used to capture logging output from all the nodes: Should an error happen or a postcondition fail, we can dump the log of each node which is invaluable to troubleshoot such failures. In general, testing systems in a black-box way emphasises the importance of good logging to provide as much context as possible should issues arise, and using quickcheck-dynamic makes no exception.
```

Comment:
I'm not sure about this. The `logger` here is not capturing the logging output. Indeed the `IOSim` monad does it and we can retrieve is using `selectTraceEventsDynamic`. So it actually is NOT required to keep this in the `StateT`, but handing the tracer to `perform`/`runModel` would be sufficient for that.